server/src/main/java/org/elasticsearch/health/node/HealthInfoCache.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health.node;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.health.node.selection.HealthNode;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Keeps track of several health statuses per node that can be used in health.
+ */
+public class HealthInfoCache implements ClusterStateListener {
+
+    private volatile ConcurrentHashMap<String, DiskHealthInfo> diskInfoByNode = new ConcurrentHashMap<>();
+
+    private HealthInfoCache() {}
+
+    public static HealthInfoCache create(ClusterService clusterService) {
+        HealthInfoCache healthInfoCache = new HealthInfoCache();
+        clusterService.addListener(healthInfoCache);
+        return healthInfoCache;
+    }
+
+    public void updateNodeHealth(String nodeId, DiskHealthInfo diskHealthInfo) {
+        diskInfoByNode.put(nodeId, diskHealthInfo);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentHealthNode = HealthNode.findHealthNode(event.state());
```

Comment:
Shall we shortcircuit this method a bit with a check for null on `currentHealthNode` and do an empty `return` in case it's `null`? 

I know we are not at risk of NPE because of the way we call `equals` below, but I'd rather have the 
nullable explicitly visible

Reply:
I wanted to take the fact that the `currentHealthNode` becoming `null` as an event to trigger the reset of the values. I think it makes sense to reset it there. For this reason I do not want to do an empty return. But I am reworking the code to make it a bit cleaner. So we can check if the readability has improved on the reworked code.

Reply:
Ignore my previous answer, I thought this comment referred to different file. 

If the `currentHealthNode` is null we still want to empty the cache, so we cannot return immediately, but I can add an explicit check that `currentHealthNode` is not `null`.